kolibri/auth/class_permissions_prototype_models.py

```diff
@@ -0,0 +1,170 @@
+            "collection_table": Collection._meta.db_table,
+        }
+
+        # list out the tables (along with aliases) that we'll need for use in the WHERE clause
+        self.tables = (
+            '"{collection_table}" AS "source_collection"',
+            '"{collection_table}" AS "target_collection"',
+            '"{role_table}" AS "target_role"',
+        )
+
+        # list out the primary role-based conditions for the WHERE clause
+        self.base_conditions = (
+            "{source_role}.collection_id = source_collection.id",
+            "{source_role}.type != 'learner'",
+            "target_collection.lft BETWEEN source_collection.lft AND source_collection.rght",
+        )
+
+    def filter_by_relationship(self, source_user, target_user=None, target_collection=None):
+        """
+        Filters a queryset of Role objects to include only those that link the source_user to the target_user
+        or target_collection, through the Collection hierarchy.
+        """
+
+        assert (target_user or target_collection) and not (target_user and target_collection),
+            "Exactly one of `target_user` and `target_collection` must be provided."
+
+        # start with the basic conditions, establishing relationship between source user, her roles, and collections
+        conditions = self.base_conditions += (
+            "{source_role}.user_id = {user_id}".format(user_id=source_user.id),
+        )
+
+        # if the target is a user, establish rules linking that user to the target collection(s) via a role
+        if target_user:
+            conditions += (
+                "target_role.collection_id = target_collection.id",
+                "target_role.type = 'learner'",
+                "target_role.user_id = {user_id}".format(user_id=target_user.id),
+            )
+
+        # if the target is a collection, link that collection to the target collection
+        if target_collection:
+            conditions += (
+                "target_collection.id = {collection_id}".format(collection_id=target_collection.id),
+            )
+
+        # execute the query to get a queryset of all roles the provided user has in relation to target user/collection
+        queryset = self.extra(
+            tables=self._format_with_table_names(tables),
+            where=self._format_with_table_names(conditions)
+        )
+
+        return queryset
+
+    def _format_with_table_names(self, items):
+        return [item.format(**self.table_names) for item in items]
+
+
+class Role(models.Model):
+
+    kind = models.CharField(max_length=50, choices=ROLE_CHOICES)
+    user = models.ForeignKey('FacilityUser')
+    collection = models.ForeignKey('Collection')
+
+    objects = RoleQuerySet.as_manager()
+
+
+class DeviceOwner(BaseUser):
+
+    permissions = IsDeviceOwner()
+
+    # ...
+
+    def role_types_in_relation_to_user(self, target_user):
```

Comment:
It's tricky. I was trying to err on the side of being explicit. There are 2 users involved: the one whose role we're checking, and the other user the role is in relation to. I tried numerous shorter names and they felt ambiguous.

Also, for `roles` vs `role_types`: this function returns a set of strings, rather than a `QuerySet` of `Role` objects, so I was trying to distinguish that fact.